scipy/optimize/optimize.py

```diff
@@ -974,14 +992,40 @@ def check_grad(func, grad, x0, *args, **kwargs):
     >>> from scipy.optimize import check_grad
     >>> check_grad(func, grad, [1.5, -1.5])
     2.9802322387695312e-08
+    >>> rng = np.random.default_rng()
+    >>> check_grad(func, grad, [1.5, -1.5], 
+    ...             random_projection=True, seed=rng)
+    2.9802322387695312e-08
 
     """
+    x0 = np.asarray(x0)
     step = kwargs.pop('epsilon', _epsilon)
+    random_projection = kwargs.pop('random_projection', False)
+    if random_projection:
+        random_state = check_random_state(kwargs.pop('seed', None))
     if kwargs:
-        raise ValueError("Unknown keyword arguments: %r" %
+        raise ValueError("Unexpected keyword arguments: %r" %
                          (list(kwargs.keys()),))
-    return sqrt(sum((grad(x0, *args) -
-                     approx_fprime(x0, func, step, *args))**2))
+    
+    def g(w, *args):
+        func, x0, v = args[0:3]
+        return func(x0 + w*v, *args[3:])
+
+    if random_projection:
+        v = rng_integers(random_state, 0, 2, size=x0.shape)
```

Comment:
Why is this using random integers instead of a random normal? In principle, anything is fine but I think the random normal is more likely to catch bugs. It's also exactly what JAX uses, which was developed by experts on automatic differentiation.

Reply:
I am using vectors(v) randomly filled with +1 and -1. I did it for stability. As we increase the dimension of input of the function under consideration, results fluctuate a lot. Possibly this is due to very small numbers multiplied with relatively larger ones (w*v). I thought results should at least behave the same way when in both, `random_projection=True` and `random_projection=False`. This doesn't happen when we use `rand`. I can use it if stability is not a priority.

Reply:
Example:

**Code**
```python
import numpy as np
def func(x):
    return x.sum()
def grad(x):
    return np.ones(x.shape)
from scipy.optimize import check_grad

rng = np.random.default_rng(seed=1234)

sizes = [32, 64, 128, 256, 512, 1024]

for size in sizes:

    x0 = np.random.normal(size=size)

    print("Dimensions in x: ", x0.shape[0])
    r = check_grad(func, grad, x0, 
                    random_projection=True, seed=rng)
    print("Error (random_projection=True): ", r)
    r = check_grad(func, grad, [1, -1])
    print("Error (random_projection=False): ", r)

    print()
```

**Output**

Using,

```python
v = random_state.normal(size=x0.shape)
```

```bash
Dimensions in x:  32
Error (random_projection=True):  3.5243703511156355e-08
Error (random_projection=False):  0.0

Dimensions in x:  64
Error (random_projection=True):  1.5745057702076792e-08
Error (random_projection=False):  0.0

Dimensions in x:  128
Error (random_projection=True):  2.17404947377986e-07 # Will fail in tests
Error (random_projection=False):  0.0

Dimensions in x:  256
Error (random_projection=True):  2.456318348720288e-07 # Will fail in tests
Error (random_projection=False):  0.0

Dimensions in x:  512
Error (random_projection=True):  1.1026638091493624e-07 # Will fail in tests
Error (random_projection=False):  0.0

Dimensions in x:  1024
Error (random_projection=True):  7.379824342734764e-07 # Will fail in tests
Error (random_projection=False):  0.0
```

Using,

```python
v = rng_integers(random_state, 0, 2, size=x0.shape)
v[v == 0] = -1
```

```bash
Dimensions in x:  32
Error (random_projection=True):  0.0
Error (random_projection=False):  0.0

Dimensions in x:  64
Error (random_projection=True):  0.0
Error (random_projection=False):  0.0

Dimensions in x:  128
Error (random_projection=True):  0.0
Error (random_projection=False):  0.0

Dimensions in x:  256
Error (random_projection=True):  0.0
Error (random_projection=False):  0.0

Dimensions in x:  512
Error (random_projection=True):  0.0
Error (random_projection=False):  0.0

Dimensions in x:  1024
Error (random_projection=True):  0.0
Error (random_projection=False):  0.0
```

Reply:
What if instead you also normalize the random normal vector to always have length 1?

Reply:
Let me check if that helps.

Reply:
When we use the following,
```python
v = random_state.normal(0, 1, size=(x0.shape))
v = v/sqrt(sum(v**2))
step = step*100 # ad-hoc, but when values in `v` are in the order of 1e-2, then step*v will come back to the scale of `_epsilon` and hence, the numbers won't be too small to produce numerical instability (which is the case if we don't do this, see another example below).
```

```bash
Dimensions in x:  32
Error (random_projection=True):  2.5898094779819303e-10
Error (random_projection=False):  0.0

Dimensions in x:  64
Error (random_projection=True):  1.6567421878832533e-09
Error (random_projection=False):  0.0

Dimensions in x:  128
Error (random_projection=True):  4.1467763944869063e-10
Error (random_projection=False):  0.0

Dimensions in x:  256
Error (random_projection=True):  8.922983463222067e-10
Error (random_projection=False):  0.0

Dimensions in x:  512
Error (random_projection=True):  1.2426024653677814e-09
Error (random_projection=False):  0.0

Dimensions in x:  1024
Error (random_projection=True):  1.1659851750778216e-08
Error (random_projection=False):  0.0
```

When we do not do `step = step*100`,
```bash
Dimensions in x:  32
Error (random_projection=True):  3.7164398225542072e-09
Error (random_projection=False):  0.0

Dimensions in x:  64
Error (random_projection=True):  2.4010250787576126e-07
Error (random_projection=False):  0.0

Dimensions in x:  128
Error (random_projection=True):  2.0571604575359004e-07
Error (random_projection=False):  0.0

Dimensions in x:  256
Error (random_projection=True):  3.9036482102350334e-08
Error (random_projection=False):  0.0

Dimensions in x:  512
Error (random_projection=True):  4.0568207415603297e-07
Error (random_projection=False):  0.0

Dimensions in x:  1024
Error (random_projection=True):  3.47541619039049e-07
Error (random_projection=False):  0.0
```

Reply:
The above results are wrong. I checked the code which I used for testing. It seems like I was not passing same input to evaluate, `random_projection=True` and `random_projection=False`. `random.normal` works fine and produces similar results. Apologies for the confusion.

Reply:
I have used random normal in the latest commit.